In [2]:
import json
import pandas as pd

In [3]:
mainfile = "result_2023_6_8.json"

In [4]:
maindata = []
with open(mainfile, 'r') as f:
    for entry in f:
        oneEntry = json.loads(entry)
        maindata.append(oneEntry)


Removing any redundancies

In [5]:
filtereddata = []
for d in maindata:
    if d.get('report') != None:
        if d.get('report').get('user_id') != None:
            filtereddata.append(d)

In [6]:
type(maindata)

list

Removing all non-GET request entries

In [7]:
getrequests = []
for d in filtereddata:
    if d.get('report').get('request_method') == 'GET':      
        getrequests.append(d)

In [8]:
print(len(maindata))
print(len(filtereddata))
print(len(getrequests))

688185
533229
300719


In [9]:
import copy

users = {}

emptydata = {
    "lastname":0,
    "firstname":0,
    "preferredname":0,
    "id":0,
    "email":0,
    "phone":0,
}


URLs = {}


user_data = {
    "URL":copy.deepcopy(emptydata),
    "body":copy.deepcopy(emptydata),
    "URLs":copy.deepcopy(URLs)
}


In [10]:
for d in getrequests:
    user_id = d.get('report').get('user_id')
    if user_id not in users:
        users[user_id] = copy.deepcopy(user_data)
        print(user_id)

d1wgiodc53k
2wpdvj73oou
zwq78tx53ib
zoumk7kg3yq
9tgpozqqg7d
s4ni7dbb7y
5tpckk3d14e
grr97iqp57r
kix7pup4zh
699mztw4z75
dknglbh1h3
p0fjtio4mr
80yfsk3rrlc
zu7pjmkeoi
4485kqex0zs
4q68v2cg2pl
l9duvpdz9m8
y9y0arr5kpm
jktgjnvw8r


In [11]:
for d in getrequests:
    user_id = d.get('report').get('user_id')
    url_leak_type = d.get('report').get('url_leak_type')
    body_leak_type = d.get('report').get('body_leak_type')
    domain = d.get('report').get('initiator_domain')
    
    
    if url_leak_type != None:
        for e in url_leak_type:
            users[user_id]["URL"][e] = users[user_id]["URL"][e] + 1
    
    if body_leak_type != None:
        for e in body_leak_type:
            users[user_id]["body"][e] = users[user_id]["body"][e] + 1
    
    if domain != None:
        if domain in users[user_id]["URLs"]:
            users[user_id]["URLs"][domain] += 1
        else:
            users[user_id]["URLs"][domain] = 1          

Just turning the above data into a nice table using tabulate

In [12]:
from tabulate import tabulate

data = []
for u in users:
    entry = []
    entry.append(u)
    for e in users.get(u).get("URL"):
        entry.append(users.get(u).get("URL").get(e))
    
    data.append(entry)

headers = ['USER ID', 'Last Name', 'First Name','Pref. Name', 'ID', 'Email', "Phone #"]
table = tabulate(data, headers=headers, tablefmt='grid')
print(table)

+-------------+-------------+--------------+--------------+------+---------+-----------+
| USER ID     |   Last Name |   First Name |   Pref. Name |   ID |   Email |   Phone # |
+=============+=============+==============+==============+======+=========+===========+
| d1wgiodc53k |          42 |            0 |            0 |    3 |      38 |         0 |
+-------------+-------------+--------------+--------------+------+---------+-----------+
| 2wpdvj73oou |           0 |            0 |            0 |    0 |       0 |         0 |
+-------------+-------------+--------------+--------------+------+---------+-----------+
| zwq78tx53ib |           4 |            0 |            0 |    0 |       4 |         0 |
+-------------+-------------+--------------+--------------+------+---------+-----------+
| zoumk7kg3yq |           0 |            0 |            5 |    0 |       0 |         0 |
+-------------+-------------+--------------+--------------+------+---------+-----------+
| 9tgpozqqg7d |      

Evidently, users "dknglbh1h3" and "zu7pjmkeoi" are outliers. Interestingly, this data doesn't quite line up with some older calculations for the number of leaks of each type, though those calculations probably didn't use the same filtering methodology I did. Everything's on the same order of magnitude, so it doesn't look like we have any problems. I've excluded those two users from the data set for the later analyses.

I'm also only showing data from the url_leak_type field, and not the body_leak_type field, even though I collected data from both. As it turns out, no personal information was found in the body_leak_type field. Possibly just encrypted, not absent, though. See below for the dictionaries.

(For what it's worth, rerunning this notebook from the top shuffles the order the USER IDs are presented in. Not sure why, but something to note down, I suppose.)

In [13]:
for u in users:
    print(u + "\t" + str(users.get(u).get("body")) )
    
for u in users:
    print(u + "\t" + str(users.get(u).get("URLs")) )

d1wgiodc53k	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
2wpdvj73oou	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
zwq78tx53ib	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
zoumk7kg3yq	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
9tgpozqqg7d	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
s4ni7dbb7y	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
5tpckk3d14e	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
grr97iqp57r	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
kix7pup4zh	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
699mztw4z75	{'lastname': 0, 'firstname': 0, 'preferredname': 0, 'id': 0, 'email': 0, 'phone': 0}
dknglbh1h3	{'lastname': 0, 'firs

In [14]:
normalizeddata = []

for d in getrequests: 
    if d.get('report').get('user_id') != "dknglbh1h3" and d.get('report').get('user_id') != "zu7pjmkeoi":
        normalizeddata.append(d)

print(len(normalizeddata))            

186308


In [15]:
leaksOnlyNormalized = []

for d in normalizeddata: 
    if d.get('report').get('initiator_domain') != None:
        leaksOnlyNormalized.append(d)
        
print(len(leaksOnlyNormalized))            

592


The number of entries was reduced by 114,411 via filtering out of the two outlying users. There are 114,021 entries in which dknglbh1h3 or zu7pjmkeoi had a last name leak. Numbers seemingly track-- those two users had their last name "leaked" with nearly every search.

Filtering out entries without an initiator domain, we find that there are only 592 entries (down from the initial >100,000 leaked data entries detected-- so evidently those users were drastically skewing the data.

In [16]:
import copy

sites = {}

leakCount = {
    "lastname":0,
    "firstname":0,
    "preferredname":0,
    "id":0,
    "email":0,
    "phone":0,
}

In [17]:
initiatorDomains = []

for d in leaksOnlyNormalized:
    domain = d.get('report').get('initiator_domain')
    if domain not in sites:
        sites[domain] = copy.deepcopy(leakCount)
        print(domain)
        initiatorDomains.append(domain)


https://www.google.com/
https://uchicago.zoom.us/
https://outlook.office.com/
https://www.youtube.com/
https://webshell.suite.office.com/
https://www.loom.com/
https://quizlet.com/
https://calendly.com/
https://scone-pa.clients6.google.com/
https://www.instagram.com/
https://ogs.google.com/
https://docs.google.com/
https://drive.google.com/
https://accounts.google.com/
https://twitter.com/
https://calendar.google.com/
https://mail.google.com/
https://www.facebook.com/
https://www.amazon.com/


The list of initiator domains is worryingly short? Don't know if this pilot dataset is just that small or if something else is going on. Maybe an oversight on my part, but might just also mean we need a bigger dataset to make any more meaningful conclusions.

In [18]:
for d in leaksOnlyNormalized:
    domain = d.get('report').get('initiator_domain')
    url_leak_type = d.get('report').get('url_leak_type')
    
    if url_leak_type != None:
        for e in url_leak_type:
            sites[domain][e] = sites[domain][e] + 1

In [19]:
data = []
for d in sites:
    entry = []
    entry.append(d)
    for e in sites.get(d):
        entry.append(sites.get(d).get(e))
    
    data.append(entry)

headers = ['DOMAIN', 'Last Name', 'First Name','Pref. Name', 'ID', 'Email', "Phone #"]
table = tabulate(data, headers=headers, tablefmt='grid')
print(table)

+---------------------------------------+-------------+--------------+--------------+------+---------+-----------+
| DOMAIN                                |   Last Name |   First Name |   Pref. Name |   ID |   Email |   Phone # |
+=======================================+=============+==============+==============+======+=========+===========+
| https://www.google.com/               |          62 |          194 |           38 |    3 |      48 |        23 |
+---------------------------------------+-------------+--------------+--------------+------+---------+-----------+
| https://uchicago.zoom.us/             |           0 |            1 |            1 |    0 |       0 |         0 |
+---------------------------------------+-------------+--------------+--------------+------+---------+-----------+
| https://outlook.office.com/           |           0 |           10 |           10 |    0 |       4 |         0 |
+---------------------------------------+-------------+--------------+----------

This has already been said, but with internal trackers, I doubt this is the full extent of what data collection is occurring

In [20]:
from urllib.parse import urlparse, urlunparse

def shorten(url):
    # Parse the URL into its components
    parsed_url = urlparse(url)

    # Reconstruct the URL with only the scheme, netloc, and fragment
    new_url = urlunparse((parsed_url.scheme, parsed_url.netloc, '', '', '', parsed_url.fragment))

    return new_url

In [21]:
domainsAndLeaks = {}


for mainDomain in initiatorDomains:
    
    leakURLs = {}
    
    for d in leaksOnlyNormalized:
        domain = d.get('report').get('initiator_domain')
        if domain == mainDomain:
            leakURL = d.get('report').get('leak_url')
            leakURL = shorten(leakURL)
        
            if leakURL not in leakURLs:
                leakURLs[leakURL] = 0
            else:
                leakURLs[leakURL] += 1
                
    domainsAndLeaks[mainDomain] = leakURLs
    
print(domainsAndLeaks["https://www.google.com/"])
        

{'https://www.google.com': 131, 'https://fonts.gstatic.com': 8, 'https://www.gstatic.com': 34, 'https://id.google.com': 2, 'https://www.google.coLEAKED_FIRSTNAME': 43, 'https://www.gstatic.coLEAKED_FIRSTNAME': 33, 'https://apis.google.coLEAKED_FIRSTNAME': 7, '': 1, 'https://encrypted-tbn0.gstatic.com': 3, 'https://adservice.LEAKED_FIRSTNAME.com': 7, 'https://ogs.google.com': 0, 'https://www.google.LEAKED_FIRSTNAME': 1, 'https://lh3.googleusercontent.com': 1}


Some weird quirks here: for example, some of the URLs seem to be broken & including "LEAKED_FIRSTNAME" in places where it clearly shouldn't be. There's also an entire URL that seems to be wiped clean (I suspect that was marked Null in the original dataset, though how there's a leak without a Leak URL probably also warrants investigation). It might be due to the package I'm using to shorten the URLs, but I suspect it's more likely a relic of the scanner that needs to be fixed: maybe it should only be looking for leaks after the top level domain? Warrants discussion-- so for now I'll leave the error intact instead of manually paving over it until the cause is found

In [22]:
for d in initiatorDomains:
    print(d)

https://www.google.com/
https://uchicago.zoom.us/
https://outlook.office.com/
https://www.youtube.com/
https://webshell.suite.office.com/
https://www.loom.com/
https://quizlet.com/
https://calendly.com/
https://scone-pa.clients6.google.com/
https://www.instagram.com/
https://ogs.google.com/
https://docs.google.com/
https://drive.google.com/
https://accounts.google.com/
https://twitter.com/
https://calendar.google.com/
https://mail.google.com/
https://www.facebook.com/
https://www.amazon.com/


In [ ]:
import os
os.system('cls')

value = input("Please request an initiator domain from the list above, or type 'exit' to close: ")

while True:
    
    os.system('cls')
    if value == "exit":
        break
    data = domainsAndLeaks[value]
    table_data = [[key, value] for key, value in data.items()]

    table = tabulate(table_data, headers=["Attribute", "Value"], tablefmt="grid")
    print(table)
    print("\n\n")
    value = input("Please request an initiator domain, or type 'exit' to close: ")
    

This is a clunky little piece of code, and in future versions it might be better to just print everything, but I find this is a good way to make it easy to compare two specific URLs and keep the view short and sweet by focusing on what we determine relevant.